### ISSUE TITLE
Getting this issue when trying to sync

### ISSUE THREAD

**Issue by jspenc72** [pouchdb/pouchdb#3592]
error JSON.stringify()ing argument: TypeError: JSON.stringify cannot serialize cyclic structures.

**Comment by nolanlawson** [comment]
@jspenc72 Can you check to make sure that the documents you are trying to save can be stringified?  E.g. can you do

```
JSON.parse(JSON.stringify(myDocument))
```

If your document is cyclic (i.e. some fields point to fields that point back to those same fields), then you cannot store that in PouchDB.

**Comment by jspenc72** [comment]
Here is what i am working with. I understand that the error arises from cyclic nature of the doc. It is displayed in that one of the timeEntrys references the _id its parent doc...


{
   "_id": "project_2014-12-28T00:08:25.430Z",
   "_rev": "362-9dafa29776ee8f69dca45021499ee80b",
   "start_date": "2014-11-21T14:08:38.093",
   "end_date": "2015-02-16T14:08:39",
   "type": "project",
   "tax_rate": 0,
   "timeline": {
       "milestones": [

           {
               "id": "milestone_2014-12-02T20:47:51.137Z"
           },
           {
               "id": "milestone_2014-12-02T20:52:16.760Z"
           },
           {
               "id": "milestone_2014-12-02T21:24:48.912Z"
           },
           {
               "id": "milestone_2014-12-10T22:16:53.273Z"
           }
       ],
       "phases": [
           {
               "title": "December 2014",
               "id": "phase_2014-12-27T06:56:54.830Z",
               "milestones": [
                   {
                       "title": "Secure funding",
                       "id": "milestone_2014-11-21T21:09:36.798Z"
                   }
               ]
           },
           {
               "title": "January 2015",
               "id": "phase_2014-12-27T07:12:04.771Z",
               "milestones": [
                   {
                       "title": "Identify Complete Process",
                       "id": "milestone_2014-11-21T21:09:21.013Z"
                   }
               ]
           }
       ]
   },
   "actual": {
       "expenseEntrys": [],
       "materialEntrys": [
     ],
       "timeEntrys": [
           {
               "id": "time_2014-12-27T20:36:43.965Z",
               "_id": "time_2014-12-27T20:27:26.036Z"
           },
           {
               "id": "time_2014-12-27T20:52:08.910Z",
               "_id": "time_2014-12-27T20:27:58.127Z",
           },
           {
               "type": "time",
               "_id": "time_2015-01-05T17:35:24.677Z",
               "_rev": "5-343d44e280534d9661e211a7b49765de",
               "project": {
                   "_id": "project_2014-12-28T00:08:25.430Z",
                   "title": "Powder Production - ARComp"
               }
           }
       ]
   },
   "budgeted": {
       "expenseEntrys": [ ]
   }
}

**Comment by jspenc72** [comment]
However, the question i do not understand is why i am able to save to pouchdb and then sync to couchdb.

**Comment by jspenc72** [comment]
Then the cyclic structure error only shows up when i am trying to sync the document to a new, empty pouch?

**Comment by jspenc72** [comment]
Additionally, excessive repeat attempts to sync the document to pouchdb will eventually succeed...

**Comment by jspenc72** [comment]
But usually this is not until all other docs in the couchdb have been synced...

**Comment by jspenc72** [comment]
I am using the latest pouchdb 3.3.1 and couchdb 1.6

**Comment by jspenc72** [comment]
@nolanlawson 
It would be nice to have some sort of graceful handling of cyclic structures. And related errors. 
Or prevent the storage of cyclic docs completely.

**Comment by jspenc72** [comment]
@nolanlawson 
Im not sure the best way to test the error but if you are interested in viewing the problem yourself I can give you access to the couchdb and let you attempt to sync?

Best,
Jesse

**Comment by nolanlawson** [comment]
Jesse: You may be running into issues because IndexedDB is somewhat lax about the objects it allows you to store. But you should not count on IndexedDB, since CouchDB/WebSQL/LevelDB all require regular JSON.

Please just remove the cyclic dependencies from your documents before storing. :)

**Comment by jspenc72** [comment]
Thank you for your time. I am removing dependencies however it might be interesting to note that the error arrises while  running pouchdb in a cordova app on iOS 8.

**Comment by jspenc72** [comment]
I am using the sqlite plugin.

**Comment by jspenc72** [comment]
Surprisingly, the error persisted after I removed the cyclic dependency I had introduced. 

It seemed related to the method of "upsertion" I was using to resolve occasional conflicts. It now appears to be resolved after recreating the couchdb by using alldocs to transfer the docs from a backup pouchdb to clear all document revisions.

**Comment by jspenc72** [comment]
I will switch to the upsert couchdb plugin.

**Comment by nickcolley** [comment]
@jspenc72 closing this for now, please reopen if you need any more help! :+1:
